### PR TITLE
fix: improve error message when can't train PQ on too small dataset

### PR DESCRIPTION
**Pull request by albertlockett**
adds a more helpful error message for https://github.com/lancedb/lancedb/issues/1465

### CODE REVIEW

File: rust/lance/src/index/vector/pq.rs [hunk 1]
```diff
@@ -408,6 +408,19 @@ pub async fn build_pq_model(
     } else {
         training_data
     };
+
```
Comment (broccoliSpicy) [inline rust/lance/src/index/vector/pq.rs:411]: this logic might be similar with what we have in `_verify_pq_sample_rate`, maybe we can modify `_verify_pq_sample_rate` to have `num_codes`

**Comment by broccoliSpicy** [comment]
the error of `https://github.com/lancedb/lancedb/issues/1465` might happened before training pq, in the `train_ivf` phase

File: rust/lance/src/index/vector/pq.rs [hunk 2]
```diff
@@ -408,6 +408,19 @@ pub async fn build_pq_model(
...
+    let num_codes = 2_usize.pow(params.num_bits as u32);
+    if training_data.len() < num_codes {
+        return Err(Error::Index {
+            message: format!(
+                "Not enough rows to train PQ. Requires {:?} rows but only {:?} available",
```
[the remark this answers is not in the document]
Comment (albertlockett) [reply rust/lance/src/index/vector/pq.rs:416]: @wjones127 added

File: rust/lance/src/index/vector/pq.rs (hunk 1 above)
Comment (albertlockett) [reply rust/lance/src/index/vector/pq.rs:411]: We have the parameter `num_bits` which can be used to control the number of codes, but using fewer bits/codes could negatively affect index recall

**Review by wjones127 — APPROVED** [review]

**Pull request merged**